### PR TITLE
Introduce STM32 I2C driver 

### PR DESCRIPTION
**Pull request by etienne-lms**

### CODE REVIEW

**Comment by etienne-lms** [comment]
note: checkpatch still complains on URL in commit message that oversize 75 chars.

File: core/drivers/stm32_i2c.c [hunk 1]
```diff
@@ -0,0 +1,1389 @@
...
+struct i2c_timing_s {
+	uint8_t scldel;
+	uint8_t sdadel;
+	uint8_t sclh;
+	uint8_t scll;
+	bool is_saved;
+};
+
+/**
+ * All these values are coming from I2C Specification, Version 6.0, 4th of
```
Comment (jbech-linaro) [inline core/drivers/stm32_i2c.c]: Can we get access to this specification? http link?

**Comment by jbech-linaro** [comment]
I'm happy to finally see an I2C driver, thanks! I've just skimmed the code. General nits, the new rule that all variables should have a well defined value (i.e., initialized in one or another way) isn't followed in this commit.

Then a more general comments/questions:
* How will this be used? From TEE core only? I cannot see any globally exported functions. I.e, I'd expect to have something like for example `core/include/i2c.h`.
* Will TA's use I2C devices in one or another way?
 ** Interfaces in that case?
 ** How to share the resource/I2C device between different TA's?
* As mentioned, I just skimmed the code and most out of curiosity, what about interrupts vs polling vs busy wait when talking to the I2C device. Seems like it is busy wait with a timeout in use. I.e., we will not "yield" to let OP-TEE do other stuff and catch data in a callback.

File: core/drivers/stm32_i2c.c [hunk 2]
```diff
@@ -0,0 +1,1383 @@
...
+/*
+ * Wait for one or more I2C_ISR bits reach their target value
+ *
+ * @hi2c: I2C handle structure
+ * @value_mask: Bit mask for the bit positions to consider in @awaited_value
+ * @awaited_value: The awaited value of I2C_ISR once masked with @value_mask
+ * @timeout_ref: Expriation timeout reference
+ * Return 0 on success or a negative value.
+ */
+static int wait_isr_bits(struct i2c_handle_s *hi2c, uint32_t value_mask,
```
Comment (etienne-lms) [inline core/drivers/stm32_i2c.c]: From an internal discussion, it appears that the description here is false and use of this function is confusing. I'll push an update to clarify.

**Comment by etienne-lms** [comment]
Rebasing since changes in **io.h**.

**Comment by etienne-lms** [comment]
Rebased, now adding few more fixup commit:
- use `vaddr_t` rather than `uintptr_t`,
- use `io_read32()` and friends,
- change few types in the driver API instead of always using `uint32_t` type.

**Comment by jforissier** [comment]
> note: checkpatch still complains on URL in commit message that oversize 75 chars.

One way to fix that is to have the URL in a `Link:` tag:

```
This is inspired from [1].

Link: [1] https://chromium.googlesource.com/chromiumos/platform/ec/+/master/include/util.h
Signed-off-by: ...
```

WIth that fixed, for commit "libutil: add UDIV_ROUND_NEAREST() macro":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/include/drivers/stm32_i2c.h [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+enum i2c_speed_e {
+	I2C_SPEED_STANDARD,	/* 100 kHz */
+	I2C_SPEED_FAST,		/* 400 kHz */
+	I2C_SPEED_FAST_PLUS,	/* 1 MHz   */
+};
+
+#define STANDARD_RATE				100000
+#define FAST_RATE				400000
+#define FAST_PLUS_RATE				1000000
```
Comment (jforissier) [inline core/include/drivers/stm32_i2c.h]: These names sound a bit too generic to me, and it looks like they could be moved them to `stm32_i2c.c`, no?

Comment (etienne-lms) [reply core/include/drivers/stm32_i2c.h]: You're right. I'll move the defines.

**Comment by etienne-lms** [comment]
Defines moved from `stm32_i2c.h` to `stm32_i2c.c`.
Fixup commits squashed and tag applied.
@jbech-linaro, any other comments?

**Comment by etienne-lms** [comment]
All tags applied. Thanks.

**Pull request merged**